common.sh

```diff
@@ -0,0 +1,265 @@
+	fmt="$1"
+	shift
+	# shellcheck disable=SC2059
+	printf "${BLUE}$fmt${RESET}\n" "$@" >&2
+}
+
+pr_error() {
+	local fmt
+	fmt="$1"
+	shift
+	# shellcheck disable=SC2059
+	printf "${RED}$fmt${RESET}\n" "$@" >&2
+}
+
+logeval() {
+	printf "${YELLOW}$ %s\n${RESET}" "$*"
+	eval "$*"
+}
+
+# Ensure error handling is enabled
+set -e
+# Nicer formatting for -x
+PS4="$ "
+
+# ==== Variable setup ====
+export UBOOT_SOURCE
+export UBOOT_OUTDIR
+export BINARIES_DIR
+export DEFCONFIG
+export UBOOT_CROSS_COMPILE
+export EMPTY_RAMDISK
+
+# Set UBOOT_SOURCE to either CI_PROJECT_DIR or the first argument to
+# the script that sourced this file
+if [ -n "$CI_PROJECT_DIR" ]; then
+	UBOOT_SOURCE="$CI_PROJECT_DIR"
+elif [ -n "$1" ]; then
+	UBOOT_SOURCE="$1"
+else
+    UBOOT_SOURCE=`pwd`
+fi
+
+assert_ci_variable() {
+	local var
+	# Set var to the value of the variable named in $1
+	eval "var=\$$1"
+	if [ -z "$var" ]; then
+		pr_error "ERROR: CI variable %s not set!" "$1"
+		return 1
+	fi
+}
+
+assert_ci_variable DEFCONFIG
+
+# U-Boot build directory
+UBOOT_OUTDIR="$UBOOT_SOURCE/.output"
+BINARIES_DIR="$UBOOT_SOURCE/binaries"
+
+# Ensure the directories exist
+mkdir -p "$UBOOT_OUTDIR"
+mkdir -p "$BINARIES_DIR"
+
+# We don't need to cross compile if building on the aarch64 runner
+UBOOT_CROSS_COMPILE=""
+if [ "$(uname -m)" != "aarch64" ]; then
+	if [ -x "$(command -v aarch64-linux-gnu-gcc)" ]; then
+		UBOOT_CROSS_COMPILE=aarch64-linux-gnu-
+	else
+		UBOOT_CROSS_COMPILE=aarch64-alpine-linux-musl-
+	fi
+fi
+
+EMPTY_RAMDISK="$UBOOT_OUTDIR/empty.gz"
+
+# ==== End variable setup ====
+
+# Make an empty gzip archive to use as the ramdisk
+printf "\0" | gzip --stdout > "$EMPTY_RAMDISK"
+
+# Make wrapper for U-Boot
+uboot_make() {
+	set -x
+	make "O=$UBOOT_OUTDIR" "CROSS_COMPILE=$UBOOT_CROSS_COMPILE" -j"$(nproc --all)" "$@"
+	set +x
+}
+
+uboot_configure() {
+	# shellcheck disable=SC2086
+	uboot_make $DEFCONFIG
+}
+
+# Create a .config and then build U-Boot
+uboot_do_build() {
+	local dtb
+	dtb="$1"
+
+	uboot_configure
+	uboot_make
+	pr_info "Finished building U-Boot!"
+}
+
+# $1: uboot binary to gzip
+# $2: variable to store gzipped binary name
+uboot_gzip() {
+	local _uboot varname
+	_uboot="$1"
+	varname="$2"
+	gzip -fk "$UBOOT_OUTDIR/$_uboot"
+	# pr_debug "varname: %s" "$varname"
+	logeval "$varname=\"${_uboot}.gz\""
+}
+
+# $1: uboot binary to gzip
+# $2: variable to store gzipped binary name
+uboot_append_dtb() {
+	local _uboot dtb varname out
+	_uboot="$1"
+	varname="$2"
+	out="${uboot}-dtb"
+
+	cat "$UBOOT_OUTDIR/$uboot" "$UBOOT_OUTDIR/u-boot.dtb" > "$UBOOT_OUTDIR/$out"
+
+	logeval "$varname=\$out"
+}
+
+build_android_boot_img_cmd() {
+	local uboot output _cmd varname
+	uboot="$1"
+	output="$2"
+	varname="$3"
+
+	_cmd="mkbootimg"
+	_cmd="$_cmd --base 0x0"
+	_cmd="$_cmd --kernel_offset 0x8000"
+	_cmd="$_cmd --pagesize 4096"
+	_cmd="$_cmd --os_patch_level 2028-09-21"
+	_cmd="$_cmd --ramdisk $EMPTY_RAMDISK"
+	_cmd="$_cmd --kernel $UBOOT_OUTDIR/$uboot -o $output"
```

Comment:
Sad to see these all hardcoded here as well, it needs rework too (in tauchgang repo as well)